### ISSUE TITLE
using tRPC in constructor or ngOnInit causes build getting stuck

### ISSUE THREAD

**Issue by goetzrobin** [analogjs/analog#468]
### Please provide the environment you discovered this bug in.

0.2.0-beta.17

### Which area/package is the issue in?

trpc

### Description

After upgrading to Angular 16.1 running `nx build trpc-app` get's stuck.
The following error is thrown:
```
ERROR TRPCClientError: fetch failed
    at TRPCClientError.from (file:///Users/robingotz/Developer/goetzrobin/analog/node_modules/.pnpm/@trpc+client@10.25.0_@trpc+server@10.25.0/node_modules/@trpc/client/dist/transformResult-403dc1d4.mjs:15:16)
....
     at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  meta: undefined,
  shape: undefined,
  data: undefined,
  [cause]: TypeError: fetch failed
      at Object.fetch (node:internal/deps/undici/undici:11457:11)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
    cause: Error: connect ECONNREFUSED 127.0.0.1:4205
        at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1494:16) {
      errno: -61,
      code: 'ECONNREFUSED',
      syscall: 'connect',
      address: '127.0.0.1',
      port: 4205
    }
  }
}
```

### Please provide the exception or error you saw

```true
ERROR TRPCClientError: fetch failed
    at TRPCClientError.from (file:///Users/robingotz/Developer/goetzrobin/analog/node_modules/.pnpm/@trpc+client@10.25.0_@trpc+server@10.25.0/node_modules/@trpc/client/dist/transformResult-403dc1d4.mjs:15:16)
....
     at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  meta: undefined,
  shape: undefined,
  data: undefined,
  [cause]: TypeError: fetch failed
      at Object.fetch (node:internal/deps/undici/undici:11457:11)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
    cause: Error: connect ECONNREFUSED 127.0.0.1:4205
        at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1494:16) {
      errno: -61,
      code: 'ECONNREFUSED',
      syscall: 'connect',
      address: '127.0.0.1',
      port: 4205
    }
  }
}
```

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

**Comment by goetzrobin** [comment]
maybe this https://github.com/nuxt/framework/pull/7805/files

**Comment by cskiwi** [comment]
> could be related to: [trpc/trpc#1847](https://github.com/trpc/trpc/discussions/1847)

Yeah, I've encountered similar errors when building.  Basically It is trying to pre-render pages that require data from the "server", however the "server" isn't running in the build process so the fetching fails. My temp workaround for local builds was having the project running when doing a build.

However this doesn't work on build machines of course, the suggestion you made might be a fix, but I don't think this is a one solution fits all type of thing.

**Comment by goetzrobin** [comment]
That’s an interesting point! I always assumed the server wasn’t ready yet and the pre-rendering happened before somehow 👀

**Comment by goetzrobin** [comment]
https://github.com/nuxt/nuxt/issues/21722

**Comment by brandonroberts** [comment]
Yes, it's possible. `globalThis.$fetch` is already available during the pre-rendering phase.

```ts
if ((globalThis as any).$fetch)) {
  globalThis.$fetch('/someUrl').then(jsonData => {
    console.log(jsonData);
  })
}
```

I've noticed this causes issues with SSR as it doesn't wait until the promise is resolved to prevent serializing the app to a string.

I thought about making an injectable `injectFetch()` that works in both browser/server/pre-rendering environments but haven't moved on it yet because then we would need a custom transfer state integration to work like HTTP client.

**Comment by goetzrobin** [comment]
Okay I think we are on the right path here. When I use the custom fetch as the trpc fetch I get it to build successfully. However, if I then try to run the production build I get the following error:
```sh
vendor-cb7a4398.js:8 ERROR TRPCClientError: Cannot read properties of undefined (reading 'raw')
    at $t.from (vendor-cb7a4398.js:10:3859)
    at vendor-cb7a4398.js:10:6714
    at u.invoke (vendor-cb7a4398.js:1:6920)
    at Object.onInvoke (vendor-cb7a4398.js:8:93745)
    at u.invoke (vendor-cb7a4398.js:1:6860)
    at J.run (vendor-cb7a4398.js:1:2250)
    at vendor-cb7a4398.js:2:570
    at u.invokeTask (vendor-cb7a4398.js:1:7545)
    at Object.onInvokeTask (vendor-cb7a4398.js:8:935
```

Which comes from right here:

```ts
function customFetch(input: RequestInfo | URL, init?: RequestInit & { method: 'GET' })  {
  return (globalThis as any).$fetch.raw(input.toString(), init)
    .catch((e: any) => {
      throw e
    })
    .then((response: any) => ({
      ...response,
      headers: response.headers,
      json: () => Promise.resolve(response._data)
    }))
}
```

So it seems like `$fetch` is available during prerender but then not available on the client of the prod build 😕

**Comment by brandonroberts** [comment]
That's correct. It's available in the prod server build and during pre-rendering. 

You'd have to check for the `globalThis.$fetch` and fallback to the built-in `fetch` if it's not available

**Comment by brandonroberts** [comment]
Yes, that's correct. The global `$fetch` is available in the prod server build and during pre-rendering.

You'd have to check for the `globalThis.$fetch` and fallback to the native `fetch` if it's not available.

**Comment by goetzrobin** [comment]
Thanks for clarifying! This prompts me to ask what $fetch is. I always thought that was just a "better" fetch provided by nitro/unjs 🤔

But seems like there's more to it?

**Comment by brandonroberts** [comment]
Yea it's better in a few ways

- It works the same in node/browser/worker
- It's more type-safe out of the box as it has the return types of the API endpoints
- It's environment-aware, so if you use it during pre-rendering, it retrieves local API/data/assets without making external network requests
- It parses the response automatically, you don't have to chain the `.then(res => res.json())`

I'm sure there's more ...

**Comment by goetzrobin** [comment]
But it is not available on the client right now, correct?

Is that what you were referring to earlier that you would like it to become available with a `injectFetch` function?

**Comment by brandonroberts** [comment]
Correct. The `injectFetch` function would handle the different environments internally. Not sure how people would feel using that instead of `HttpClient` in some cases but it might be limited in usage.

**Comment by brandonroberts** [comment]
This has been released in version `0.2.3`
